active/0000-engines.md

```diff
@@ -0,0 +1,202 @@
+- Start Date: 2014-10-24
+- RFC PR:
+- Ember Issue:
+
+# Summary
+
+Engines allow multiple logical applications to be composed together into
+a single application from the user's perspective. They also allow addons
+to provide encapsulated functionality normally associated with apps,
+such as routes and templates that integrate seamlessly with the primary
+app.
+
+# Motivation
+
+Large companies are increasingly adopting Ember.js to power their entire
+product lines. Often, this means separate teams (sometimes distributed
+around the world) working on the same app. Typically, responsibility is
+shared by dividing the application into one or more "sections". How this
+division is actually implemented varies from team to team.
+
+Sometimes, each "section" will be a completely separate Ember app, with
+a shared navigation bar allowing users to switch between each app. This
+allows teams to work quickly without stepping on each others' toes, but
+switching apps feels slow (especially compared to the normally speedy
+route transitions in Ember) because the entire page must be thrown out,
+then an entirely new set of the same assets downloaded and parsed.
+Additionally, code sharing is largely accomplished via copy-and-paste.
+
+Other times, the separation is enforced socially, with each team
+claiming a section of the same app in the same repository.
+Unfortunately, this approach leads to frequent conflicts around shared
+resources, and feedback from tests gets slower and slower as test suites
+grow in size.
+
+Engines allow each section of an application to be developed in
+isolation, including separate test suites and naming conventions. These
+isolated units can be composed together into a single app from the
+perspective of the user.
+
+Additionally, engines allow addons to augment applications with
+off-the-shelf fully-functional UI. For example, imagine an
+authentication library that offers a prefab login screen that
+application developers can opt in to, dramatically increasing the speed
+at which prototypes can be developed.
+
+Or imagine a product like Discourse allowing you to mount their Ember
+app at `/forums/`, so the user needn't leave your app to read your
+discussion forum.
+
+# Detailed Design
+
+Your app can gain access to an engine in one of two ways:
+
+1. By placing an Ember CLI app inside another Ember CLI app's `engines/`
+   directory (a sibling of `app/`).
+2. By installing an Ember CLI addon that contains an `app` directory.
```

Comment:
Do you want to consider loading engines at runtime? I know some apps do something like

 1. map `/admin/*` to `route:load_admin` statically in the core app
 1. `route:load_admin` fetches the JS, CSS, and (compiled) templates for the Admin Engine
 1. `route:load_admin` defines new routes for `admin/x`, `admin/y`, etc. Because they're more specific than `admin/*`, they have higher routing priority.
 1. `route:load_admin` restarts the transition.